### PR TITLE
Find max version of an array of versions

### PR DESCRIPTION
**Pull request by oltionchampari**
Closes phovea/generator-phovea#334, phovea/generator-phovea#182

**The issue was:** 
When no intersection between the ranges is found, the **max** of these versions is calculated. Before it is calculated the method `semver.coerce(version)`  is applied in order to correctly format the versions and avoid errors.

**But:**
* `semver.coerce('4.2.0-beta.0')--> '4.2.0'`
* `semver.coerce('~4.2.0')--> '4.2.0'`

Thus, we were finding the wrong max version, instead of `'4.2.0-beta.0'` we get `'4.2.0'`.

### Summary of changes

- Added a function to calculate the max of an array of versions. 
Semver  or any other packages provides no such functionality when it comes to ranges(from what i saw).

- Added tests .

### CODE REVIEW

File: utils/version.js [hunk 1]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+  if (semver.gt(semver.coerce(parsedTildeRange.major), semver.coerce(parsedCaretRange.major))) {
+    return tildeRange;
+  } else {  // in any other case the caret range is greater(has a greater upper domain)
+    return caretRange;
+  }
+}
+
+/**
+ * Remove caret or tilde and format version
+ * @param {string} r
```
Comment (thinkh) [inline utils/version.js]: What is `r`?

File: utils/version.js [hunk 2]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+    return semver.gtr(maxNonRangeVersion, maxRange) ? maxNonRangeVersion : maxRange; // check maxNonRangeVersion is greater than all the versions possible in the range.
+  }
+
+  return maxNonRangeVersion;
+}
+
+/**
+ * Find max between a tilde range and a caret range
+ * @param {string} tildeRange Example value:'~2.0.3'
+ * @param {string} caretRange Example value'^3.5.6'
```
Comment (thinkh) [inline utils/version.js]: Spaces and punctuation.

File: utils/version.js [hunk 3]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+ */
+function findMaxRange(tildeRange, caretRange) {
+  const parsedCaretRange = semver.coerce(caretRange); // create a semver object from tag to access `major` and `version` properties
+  const parsedTildeRange = semver.coerce(tildeRange);
+  // tilde range can only be greater than caret range when its major is greater i.e '~3.5.6' > '^2.9.0'
+  if (semver.gt(semver.coerce(parsedTildeRange.major), semver.coerce(parsedCaretRange.major))) {
+    return tildeRange;
+  } else {  // in any other case the caret range is greater(has a greater upper domain)
```
Comment (thinkh) [inline utils/version.js]: Spaces and punctuation.

File: utils/version.js [hunk 4]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+
+/**
+ * Find max between a tilde range and a caret range
+ * @param {string} tildeRange Example value:'~2.0.3'
+ * @param {string} caretRange Example value'^3.5.6'
+ */
+function findMaxRange(tildeRange, caretRange) {
+  const parsedCaretRange = semver.coerce(caretRange); // create a semver object from tag to access `major` and `version` properties
+  const parsedTildeRange = semver.coerce(tildeRange);
+  // tilde range can only be greater than caret range when its major is greater i.e '~3.5.6' > '^2.9.0'
```
Comment (thinkh) [inline utils/version.js]: Spaces and punctuation.

File: utils/version.js [hunk 5]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+ * Remove caret or tilde and format version
+ * @param {string} r
+ */
+function removeRangeTag(range) {
+  return semver.prerelease(semver.minVersion(range)) ? semver.minVersion(range) : semver.coerce(range).version
+}
+
+/**
+ * Finds the max caret range of an array of caret ranges
+ * @param {string[]} versions Possible values:`['^2.3.4',^3.0.0]`
```
Comment (thinkh) [inline utils/version.js]: Spaces and punctuation.

File: test/utils-version.test.js [hunk 6]
```diff
@@ -212,3 +212,64 @@ describe('find highest version from list', () => {
...
+    const versions = ['4.2.0-alpha.1', '4.2.0-beta.0', '0.1.0', '1.0.0', '2.1.0', '2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('4.2.0-beta.0');
+  });
+
+  it('works  for arrays with prerelease and tilde ranges ', () => {
+    const versions = ['4.2.0-alpha.1', '4.2.0-beta.0', '0.1.0', '~4.2.0', '2.1.0', '~2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('~4.2.0');
+  });
+
+  it('works for arrays with prerelease and tilde and caret ranges ', () => {
```
Comment (thinkh) [inline test/utils-version.test.js]: Space

File: test/utils-version.test.js [hunk 7]
```diff
@@ -212,3 +212,64 @@ describe('find highest version from list', () => {
...
+    const versions = ['0.0.5', '0.1.0', '1.0.0', '2.1.0', '2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('2.2.0');
+  });
+
+  it('works version for arrays with prerelease versions', () => {
+    const versions = ['4.2.0-alpha.1', '4.2.0-beta.0', '0.1.0', '1.0.0', '2.1.0', '2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('4.2.0-beta.0');
+  });
+
+  it('works  for arrays with prerelease and tilde ranges ', () => {
```
Comment (thinkh) [inline test/utils-version.test.js]: Space

File: test/utils-version.test.js [hunk 8]
```diff
@@ -212,3 +212,64 @@ describe('find highest version from list', () => {
...
+    const versions = ['4.2.0-alpha.1', '4.2.0-beta.0', '0.1.0', '~4.2.0', '2.1.0', '~2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('~4.2.0');
+  });
+
+  it('works for arrays with prerelease and tilde and caret ranges ', () => {
+    const versions = ['4.2.0-alpha.1', '4.2.0-beta.0', '^4.2.0', '~4.2.0', '2.1.0', '~2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('^4.2.0');
+  });
+
+  it('works for "^4.2.0" "~4.8.0" ', () => {
```
Comment (thinkh) [inline test/utils-version.test.js]: You don't check for `~4.8.0` in the version array...

File: test/utils-version.test.js [hunk 9]
```diff
@@ -212,3 +212,64 @@ describe('find highest version from list', () => {
...
+    const versions = ['4.2.0-alpha.1', '4.2.0-beta.0', '^4.2.0', '~4.2.0', '2.1.0', '~2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('^4.2.0');
+  });
+
+  it('works for  caret prerelease ranges ', () => {
```
Comment (thinkh) [inline test/utils-version.test.js]: Spaces

File: test/utils-version.test.js [hunk 10]
```diff
@@ -212,3 +212,64 @@ describe('find highest version from list', () => {
...
+    const versions = ['^4.2.0-alpha.1', '^4.2.0-beta.0', '^4.1.0', '2.1.0', '~2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('^4.2.0-beta.0');
+  });
+
+  it('works  for tilde prerelease ranges ', () => {
```
Comment (thinkh) [inline test/utils-version.test.js]: Spaces

File: test/utils-version.test.js [hunk 11]
```diff
@@ -212,3 +212,64 @@ describe('find highest version from list', () => {
...
+    const versions = ['~4.2.0-alpha.1', '~4.2.0-beta.0', '~4.1.0', '2.1.0', '~2.2.0'];
+    expect(version.findMaxVersion(versions)).toBe('~4.2.0-beta.0');
+  });
+
+  it('works for tilde and caret prerelease ranges ', () => {
```
Comment (thinkh) [inline test/utils-version.test.js]: Space

File: utils/version.js [hunk 12]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
 };
 
+/**
+ * Finds the max version of an array of versions. i.e `['1.2.3-alpha.1', '4.0.1-beta.0', '^3.8.0', '~4.0.0', '^2.8.0', '^4.0.0', '4.0.1-rc.0']`
+ * Can handle range(caret ,tilde) and prerelease versions
+ * @param {stringArray} versions Array of semver versions even range versions
+ */
+function findMaxVersion(versions) {
+  const nonRangeVersions = versions.filter((v) => !hasRangeVersionTag(v)).map((v) => semver.prerelease(v) ? v : semver.coerce(v).version); // Filter out versions with `~` and `^`
+  // Sort versions and get max /semver.rcomapre() fails when you try to compare ranges i.e `~2.0.0`
```
Comment (thinkh) [inline utils/version.js]: Missing space and comma.

File: utils/version.js [hunk 13]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+  const caretRanges = versions.filter((v) => v.startsWith('^'));
+  if (caretRanges) {
+    const parsedTags = caretRanges.map((r) => removeRangeTag(r))
+    const max = parsedTags.sort(semver.rcompare)[0] || caretRanges[caretRanges.length - 1];
+    return caretRanges.find((r) => semver.eq(removeRangeTag(r), max));
+  }
+}
+
+/**
+ * Finds the  max tilde range of an array of tilde ranges
```
Comment (thinkh) [inline utils/version.js]: Spaces and punctuation.

File: utils/version.js [hunk 14]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+ * @param {string[]} versions Possible values: `['~2.3.4',~3.0.0]`
```
Comment (thinkh) [inline utils/version.js]: `@return` is missing

File: utils/version.js [hunk 15]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+}
+
+/**
+ * Finds the max caret range of an array of caret ranges
+ * @param {string[]} versions Possible values:`['^2.3.4',^3.0.0]`
+ */
+function findMaxCaretRange(versions) {
+  const caretRanges = versions.filter((v) => v.startsWith('^'));
+  if (caretRanges) {
+    const parsedTags = caretRanges.map((r) => removeRangeTag(r))
```
Comment (thinkh) [inline utils/version.js]: Missing semicolon

File: utils/version.js [hunk 16]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+}
+
+/**
+ * Finds the  max tilde range of an array of tilde ranges
+ * @param {string[]} versions Possible values: `['~2.3.4',~3.0.0]`
+ */
+function findMaxTildeRange(versions) {
+  const tildeRanges = versions.filter((v) => v.startsWith('~'));
+  if (tildeRanges) {
+    const parsedTags = tildeRanges.map((r) => removeRangeTag(r))
```
Comment (thinkh) [inline utils/version.js]: Missing semicolon

**Comment by thinkh** [comment]
@oltionchampari Please add a release label to this PR.

File: utils/version.js [hunk 17]
```diff
@@ -20,12 +20,89 @@ module.exports.mergeVersions = (name, versions) => {
...
+    const max = parsedTags.sort(semver.rcompare)[0] || tildeRanges[tildeRanges.length - 1];
+    return tildeRanges.find((r) => semver.eq(removeRangeTag(r), max));
+  }
+}
+
+/**
+ *Check if version is a caret range or a tilde range
+ * @param {string} version Possible values: `^=2.0.0`, `~3.5.6`, `3.4.5`
```
[the remark this answers is not in the document]
Comment (oltionchampari) [reply utils/version.js]: A string .

**Comment by oltionchampari** [comment]
@thinkh
 Thank you for the feedback. I need to be more more careful when commenting code.

**Pull request merged**